sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceImplTest.java

```diff
@@ -126,6 +143,538 @@ public void testRead() throws IOException {
+    expectedSecondRangeRows.add(FlatRow.newBuilder().withRowKey(ByteString.copyFromUtf8("c")).build());
+
+    for (int i = 1; i < MINI_BATCH_ROW_LIMIT; i++) {
+      expectedFirstRangeRows.add(FlatRow.newBuilder().withRowKey(ByteString.copyFromUtf8("b" + i)).build());
+      expectedSecondRangeRows.add(
+          FlatRow.newBuilder().withRowKey(ByteString.copyFromUtf8("d" + i)).build());
+    }
+
+    ByteKey firstStart = ByteKey.copyFrom("a".getBytes(StandardCharsets.UTF_8));
+    ByteKey sharedKeyEnd = ByteKey.copyFrom("c".getBytes(StandardCharsets.UTF_8));
+    ByteKey secondEnd = ByteKey.copyFrom("e".getBytes(StandardCharsets.UTF_8));
+
+    when(mockBigtableSource.getRanges())
+        .thenReturn(
+            Arrays.asList(
+                ByteKeyRange.of(firstStart, sharedKeyEnd),
+                ByteKeyRange.of(sharedKeyEnd, secondEnd)));
+    when(mockBigtableSource.getTableId()).thenReturn(StaticValueProvider.of(TABLE_ID));
+    when(mockBigtableDataClient.readFlatRows(any(ReadRowsRequest.class), any()))
+        .thenAnswer(new Answer<ScanHandler>() {
+          @Override
+          public ScanHandler answer(InvocationOnMock invocationOnMock) throws Throwable {
+            StreamObserver<FlatRow> flatRowObserver = invocationOnMock.getArgument(1);
+            new Thread() {
+              @Override
+              public void run() {
+                for (int i = 0; i < MINI_BATCH_ROW_LIMIT; i++) {
+                  flatRowObserver.onNext(expectedFirstRangeRows.get(i));
+                }
+                flatRowObserver.onCompleted();
+              }
+            }.start();
+
+            return scanHandler;
+          }
+        }).thenAnswer(new Answer<ScanHandler>() {
+          @Override
+          public ScanHandler answer(InvocationOnMock invocationOnMock) throws Throwable {
+            StreamObserver<FlatRow> flatRowObserver = invocationOnMock.getArgument(1);
+            new Thread() {
+              @Override
+              public void run() {
+                for (int i = 0; i < MINI_BATCH_ROW_LIMIT; i++) {
+                  flatRowObserver.onNext(expectedSecondRangeRows.get(i));
+                }
+                flatRowObserver.onCompleted();
+              }
+            }.start();
+
+            return scanHandler;
+          }
+        }).thenAnswer(new Answer<ScanHandler>() {
+          @Override
+          public ScanHandler answer(InvocationOnMock invocationOnMock) throws Throwable {
+            StreamObserver<FlatRow> flatRowObserver = invocationOnMock.getArgument(1);
+            flatRowObserver.onCompleted();
+            return scanHandler;
+          }
+        });
+
+    BigtableService.Reader underTest =
+        new BigtableServiceImpl.BigtableMiniBatchReaderImpl(mockSession, mockBigtableSource);
+
+    underTest.start();
+    Assert.assertEquals(FlatRowConverter.convert(expectedFirstRangeRows.get(0)), underTest.getCurrentRow());
+    for (int i = 1; i < MINI_BATCH_ROW_LIMIT; i++) {
```

Comment:
nit for(int i=0; i < MINI_BATCH_ROW_LIMIT - 1; i++) to make it explicit that you are advance n-1 times